tsl/src/continuous_aggs/materialize.c

```diff
@@ -660,6 +671,75 @@ scan_take_invalidation_tuple(TupleInfo *ti, void *data)
 	return SCAN_CONTINUE;
 }
 
+static ScanTupleResult
+invalidation_threshold_htid_found(TupleInfo *tinfo, void *data)
+{
+	if (tinfo->lockresult != TM_Ok)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+				 errmsg("could not acquire lock for invalidation threshold row %d",
+						tinfo->lockresult),
+				 errhint("Retry the operation again")));
+	}
+	return SCAN_DONE;
+}
+
+/* lock row corresponding to hypertable id in
+ * continuous_aggs_invalidation_threshold table in AccessExclusive mode,
+ * block till lock is acquired.
+ */
+static void
+lock_invalidation_threshold_hypertable_row(int32 raw_hypertable_id)
+{
+	ScanTupLock scantuplock = {
+		.waitpolicy = LockWaitBlock,
+		.lockmode = LockTupleExclusive,
+	};
+	Relation invalidation_threshold_table_relation;
+	Catalog *catalog = ts_catalog_get();
+	ScanKeyData scankey[1];
+	int retcnt = 0;
+
+	ScanKeyInit(&scankey[0],
+				Anum_continuous_aggs_invalidation_threshold_pkey_hypertable_id,
+				BTEqualStrategyNumber,
+				F_INT4EQ,
+				Int32GetDatum(raw_hypertable_id));
+
+	/* lock table in AccessShare mode and the row with AccessExclusive */
+	ScannerCtx scanctx = { .table = catalog_get_table_id(catalog,
+														 CONTINUOUS_AGGS_INVALIDATION_THRESHOLD),
+						   .index = catalog_get_index(catalog,
+													  CONTINUOUS_AGGS_INVALIDATION_THRESHOLD,
+													  CONTINUOUS_AGGS_INVALIDATION_THRESHOLD_PKEY),
+						   .nkeys = 1,
+						   .scankey = scankey,
+						   .limit = 1,
+						   .tuple_found = invalidation_threshold_htid_found,
+						   .lockmode = AccessShareLock,
+						   .scandirection = ForwardScanDirection,
+						   .result_mctx = CurrentMemoryContext,
+						   .tuplock = &scantuplock };
+	/* Acquire lock explicitly.
+	 * the catalog scanning code , releases it at end of the call
+	 */
+	invalidation_threshold_table_relation =
+		table_open(catalog_get_table_id(catalog, CONTINUOUS_AGGS_INVALIDATION_THRESHOLD),
+				   AccessShareLock);
+	retcnt = ts_scanner_scan(&scanctx);
+	if (retcnt > 1)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+				 errmsg("unexpected return value %d in lock_invalidation_threhold_hypertable_row",
```

Comment:
Using the function name in error messages is redundant as elog already tracks this and the function and line number can be shown if the right log configuration is set. I think it is better to formulate a good error message to describe the error, e.g.: "duplicate catalog entry for hypertable \"%s\"". As a last resort, use the `__func__` macro as that also handles function name changes.